pkg/controller/traffic/pod_label_shifter.go

```diff
@@ -18,159 +18,168 @@ import (
 		appName:               appName,
+		releaseName:           releaseName,
 		namespace:             namespace,
-		serviceSelector:       labels.Set(serviceSelector).AsSelector(),
 		clusterReleaseWeights: weights,
 	}, nil
 }
 
 func (p *podLabelShifter) SyncCluster(
 	cluster string,
-	release string,
 	clientset kubernetes.Interface,
 	informerFactory kubeinformers.SharedInformerFactory,
 ) (uint32, error) {
-	releaseWeights, ok := p.clusterReleaseWeights[cluster]
+	releaseTargetWeights, ok := p.clusterReleaseWeights[cluster]
 	if !ok {
 		return 0, shippererrors.NewMissingTrafficWeightsForClusterError(
 			p.namespace, p.appName, cluster)
 	}
 
-	appSelector := labels.Set{shipper.AppLabel: p.appName}.AsSelector()
-	releaseSelector := labels.Set{shipper.ReleaseLabel: release}.AsSelector()
-	trafficSelector, err := p.getTrafficSelector(cluster, informerFactory)
-	if err != nil {
-		return 0, err
-	}
-
 	podLister := informerFactory.Core().V1().Pods().Lister().Pods(p.namespace)
 	podGVK := corev1.SchemeGroupVersion.WithKind("Pod")
 
+	appSelector := labels.Set{shipper.AppLabel: p.appName}.AsSelector()
 	appPods, err := podLister.List(appSelector)
 	if err != nil {
 		return 0, shippererrors.NewKubeclientListError(
 			podGVK, p.namespace, appSelector, err)
 	}
 
-	releasePods, err := podLister.List(releaseSelector)
+	releaseSelector := labels.Set{shipper.ReleaseLabel: p.releaseName}.AsSelector()
+
+	svc, err := p.getService(informerFactory)
 	if err != nil {
-		return 0, shippererrors.NewKubeclientListError(
-			podGVK, p.namespace, releaseSelector, err)
+		return 0, err
+	} else if svc.Spec.Selector == nil {
+		return 0, shippererrors.NewTargetClusterServiceMissesSelectorError(p.namespace, svc.Name)
 	}
 
-	var totalWeight uint32 = 0
-	for _, weight := range releaseWeights {
-		totalWeight += weight
+	trafficSelector := labels.Set(svc.Spec.Selector).AsSelector()
+
+	endpoints, err := informerFactory.Core().V1().Endpoints().Lister().
+		Endpoints(svc.Namespace).Get(svc.Name)
+	if err != nil {
+		return 0, shippererrors.NewKubeclientGetError(svc.Namespace, svc.Name, err).
+			WithCoreV1Kind("Endpoints")
 	}
 
-	targetWeight := releaseWeights[release]
+	status := NewTrafficShiftingStatus(appPods, endpoints, trafficSelector, releaseSelector)
+
+	var totalTargetWeight uint32 = 0
+	for _, weight := range releaseTargetWeights {
+		totalTargetWeight += weight
+	}
+
+	releaseTargetWeight := releaseTargetWeights[p.releaseName]
 	targetPods := calculateReleasePodTarget(
-		len(releasePods), targetWeight, len(appPods), totalWeight)
-
-	var trafficPods []*corev1.Pod
-	var idlePods []*corev1.Pod
-	for _, pod := range releasePods {
-		if getsTraffic(pod, trafficSelector) {
-			trafficPods = append(trafficPods, pod)
-			continue
+		status.ReleasePods, releaseTargetWeight, len(appPods), totalTargetWeight)
+
+	// TODO(jgreff): we need to be able to tell, from the outside, that all
+	// pods have been labeled, but some may not be ready yet.
```

Comment:
Do you mean some kind of annotations?

Reply:
No, I mean what was achieved in https://github.com/bookingcom/shipper/pull/245/commits/a9e5162b0ac975863e9a36b7062c526bb3685d27 -- basically, returning the trafficStatus object so the controller can use it to report things better.